NIFI-14436 TestRunner ensures properties with unsatisfied dependencies are not used

EndzeitBegins opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-14436](https://issues.apache.org/jira/browse/NIFI-14436)

Adjusts the `TestRunner` to ensure properties whose dependencies are not satisfied are not used. See [NIFI-14400](https://issues.apache.org/jira/browse/NIFI-14400) for more background on the motivation behind this change.

~~The added behaviour is opt-out using `runner.setProhibitUseOfPropertiesWithUnsatisfiedDependencies(false)`. There are several existing components which do not conform to the new check. These are to be addressed in separate PRs under NIFI-14437 or a sub-task thereof.~~
The behaviour cannot be disabled. All components that did not pass the added check were refactored accordingly. 

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

pvillard31 commented:
Hey @EndzeitBegins - thanks for putting this together!

While I understand the multi steps approach, I feel like we should not even allow the option to opt-out. I looked at some of the cases where the problem is exhibited and in those cases it would not trigger a bug (ie. we do try to get the value of the property even if the dependency is not satisfied but we later check the dependency or if the returned value is null), it's more about having a better code structure.

I'm happy to collaborate with you to fix the places where we have the failure if you'd like to divide and conquer. Let me know.

exceptionfactory reviewed (commented):
I agree with @pvillard31 that the TestRunner should not have the option to opt out of the expected behavior.

Although I agree that there is value in general with making incremental changes, I think this is a case where it is better to make all of the changes together.

EndzeitBegins commented:
Alright. Then I'll continuously append commits addressing the different components to this PR instead. 

That way, changes should still be relatively easy to track while all work is kept under a single PR and may be squashed before merge if desired.
Sounds good to both of you @pvillard31 & @exceptionfactory?

exceptionfactory commented:
> Alright. Then I'll continuously append commits addressing the different components to this PR instead. In the meantime I'll mark the PR as draft.
> 
> That way, changes should still be relatively easy to track while all work is kept under a single PR and may be squashed before merge if desired. Sounds good to both of you @pvillard31 & @exceptionfactory?

That sounds like a good plan, thanks @EndzeitBegins!

EndzeitBegins commented:
I refactored all components where the added check failed and removed the opt-out functionality.

Most changes were straightforward, primarily involving conditional property checks or adjustments to the execution path. In rare instances, tests revealed that the processor relied on a specific property (or its default value) even when the dependency constraint was unmet. In these cases, I removed the constraint entirely.

I have rebased these changes onto the latest `main` branch to resolve merge conflicts. While system tests encountered issues yesterday due to changes merged into `main`, a subsequent rebase appears to have resolved these problems.

I'd appreciate it if you could take a look, @pvillard31 & @exceptionfactory.

exceptionfactory reviewed (approved):
Thanks for the extensive work on making the changes to align expected behavior @EndzeitBegins. The changes looks good.

Any additional feedback @pvillard31?

pvillard31 reviewed (approved):
Thanks for this work @EndzeitBegins - I'm a +1
